cmd/recv.go

```diff
@@ -190,52 +189,13 @@ func recvAction(cmd *cobra.Command, args []string) {
 				}
 
 				tmpFile.Seek(0, io.SeekStart)
-				zr, err := zip.NewReader(tmpFile, int64(n))
-				if err != nil {
-					bail("Read zip error: %s", err)
-				}
 
-				for _, zf := range zr.File {
-					p, err := filepath.Abs(filepath.Join(dirName, zf.Name))
-					if err != nil {
-						bail("Failes to calculate file path ABS: %s", err)
-					}
-
-					if !strings.HasPrefix(p, dirName) {
-						bail("Dangerous filename detected: %s", zf.Name)
-					}
-
-					rc, err := zf.Open()
-					if err != nil {
-						bail("Failed to open file in zip: %s %s", zf.Name, err)
-					}
-
-					dir := filepath.Dir(p)
-					err = os.MkdirAll(dir, 0777)
-					if err != nil {
-						bail("Failed to mkdirall %s: %s", dir, err)
-					}
-
-					f, err := os.Create(p)
-					if err != nil {
-						bail("Failed to open %s: %s", p, err)
-					}
-
-					_, err = io.Copy(f, rc)
-					if err != nil {
-						bail("Failed to write to %s: %s", p, err)
-					}
-
-					err = f.Close()
-					if err != nil {
-						bail("Error closing %s: %s", p, err)
-					}
-
-					rc.Close()
```

Comment:
@psanford This was what I meant. The two close calls are not deferred, but are instead in the end of the for-loop. If anything after `zf.Open()` returns an error, it is going to exit without closing that reader. More importantly, if the `io.Copy()` call fails, the file is not going to be closed. This is what I improved upon by deferring the closings and thus also making sure to handle the errors.

Reply:
Thanks for clarifying. 

If `io.Copy` fails the process exits via `bail()`. All file handles are closed by the OS when the process exits. 

`zf` is opened for reading, not writing. The article you linked to was specifically about calling `Close()` on files opened for writing.

Reply:
> If io.Copy fails the process exits via bail(). All file handles are closed by the OS when the process exits.

Yeah, sure but I still think it’s better to have the code make sure to clean up after itself first instead of having to rely on the operating system to do it all. With potential differences between operating systems, I think this is a much more elegant solution. Especially if the code is used in a situation where the program doesn’t close directly after the extraction (potentially if a user uses the example in his own project).

> zf is opened for reading, not writing. The article you linked to was specifically about calling Close() on files opened for writing.

Indeed, I know, but I figured that it can’t hurt to handle the error anyway.

Reply:
Given that this isn't addressing any actual, concrete problems, I'm inclined to not take this refactoring.

Reply:
Oh, come on now. It moves everything out to readable and well-structured files and also adds the code to the examples. I think that is significant value-add if anything.

Reply:
I mean, adding two defers for better error handling can't make anything worse. At this point, that is pretty much the only thing that has changed from the initial zip implementation (apart from the os.OpenFile specifying to only support writes).

Reply:
Sorry, but I don't think you have made the code more readable. 
- The new implementation is longer than the original. 
- It depends on named return values and mixes naked returns and explicit returns. I personally find this to be harder to reason about and it is generally discouraged within the go community.
- The new code has less specific error messages than the original.
- The use of defers do not actually simplify anything. Defers are good when they reduce the amount of code you have to write for function cleanup. Since the original code simply exited the process, it did not need to worry that.

Reply:
Alright then. Would you be happy if I moved the code back to the original file and removed the defers?

Reply:
What code changes are left at that point?

Reply:
Well, for starters we have the dire receive example. Apart from that, it’s mostly just the OpenFile being wrote only with the permissions from the ziped file and handling the error when closing the zip reader (without defer).

Reply:
Sure, revert the refactoring and keep the other changes.

Reply:
Should be resolved now :slightly_smiling_face: